File: problem_solving/dailytemp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <stack>

using namespace std;

vector<int> dailyTemperatures(vector<int>& temperatures) {
    stack<int> st;
    vector<int> res(temperatures.size());
    for (int i=0; i<temperatures.size(); i++){
        
        while (!st.empty() && temperatures[st.top()] < temperatures[i]){
            int last_index = st.top();
            res[last_index] = i - last_index;
            st.pop();
        }
        
        st.push(i);
    }
    return res;
}

int main() {
    
    vector<int> in = {73, 74, 75, 71, 69, 72, 76, 73};
    for (int x: dailyTemperatures(in)){
        cout << x << " ";
    }
    
    
    return 0;
}

File: problem_solving/dailytemp.py
def dailyTemperatures(temps):
        results = [0] * len(temps)
        stack = []
        # UPVOTE !
        for i, temp in enumerate(temps):
            while stack and temps[stack[-1]] < temp:
                index = stack.pop()
                results[index] = i - index
            stack.append(i)

        return results

print(dailyTemperatures([73, 74, 75, 71, 69, 72, 76, 73]))


class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        # Observation: We only need the values greater than the current value.
        # Monotonic stack can be an effective data structure to use in this case.
        # In the end we need to check the remaining values.
        # The best way to do it is to append 0 at the end of the heights array(unique approach).
        # Time comp.-> O(N), Space comp.-> O(N)
        heights.append(0)
        stack = []
        res = 0
        for i in range(len(heights)):
            index = i
            while stack and heights[i] < stack[-1][0]:
                value, index = stack.pop()
                area = value * (i - index)
                res = max(res, area)
            stack.append([heights[i], index])
        return res
